crates/ethereum/consensus/src/lib.rs

```diff
@@ -241,10 +247,13 @@ where
             // order, making those checks obsolete.
 
             // Check if timestamp is in the future. Clock can drift but this can be consensus issue.
+            #[cfg(feature = "std")]
             let present_timestamp =
                 SystemTime::now().duration_since(SystemTime::UNIX_EPOCH).unwrap().as_secs();
-
-            if header.timestamp() > present_timestamp + ALLOWED_FUTURE_BLOCK_TIME_SECONDS {
+            #[cfg(feature = "std")]
+            if header.timestamp() >
+                present_timestamp + alloy_eips::merge::ALLOWED_FUTURE_BLOCK_TIME_SECONDS
+            {
                 return Err(ConsensusError::TimestampIsInFuture {
                     timestamp: header.timestamp(),
                     present_timestamp,
```

Comment:
Pre-merge this is not safe, post-merge we need to only have a check that the this block comes later than the parent. The beacon chain will check whether the block is too far into the future.

See the corresponding code in geth [here](https://github.com/ethereum/go-ethereum/blob/bf6da20012f63573fff4dad19634c5bf5dbef964/consensus/beacon/consensus.go#L221)

Reply:
Question: If the beacon chain has a bug or it doesn't do this check properly or perhaps `ALLOWED_FUTURE_BLOCK_TIME_SECONDS` does not correspond to the beacon chain constant then it seems like geth would accept the block (because it is in the future) but reth would not?

Reply:
The code that corresponds to the geth comment is in `validate_header_against_parent` which calls `validate_against_parent_timestamp`

Reply:
On the "not safe" note: If we were to verify the above function in a zkEVM, we would not b able to trust the time that the prover was using. They could lie and say that the time is far into the future and that this block is therefore invalid.

It's not possible for a verifier to check this because they don't know when the prover received the block and therefore can't tell if they were lying.

This is really only unsafe, on a live pre-merge chain. Since we know all of the valid blocks in the pre-merge chain, this is okay though.

The solution for live pow chains, is to have this check always pass, ie its a noop. And the verifier will check the time themselves when they receive the block header.

Reply:
should we just remove this check then?

Reply:
I guess if you want to support pre-merge blocks in std then its still useful